WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -0,0 +1,191 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Blockchain.Keys;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Tor.Http;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using WalletWasabi.WabiSabi.Crypto;
+using WalletWasabi.Wallets;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class CoinJoinClient : BackgroundService, IDisposable
+	{
+		private bool _disposedValue;
+		private CredentialPool AmountCredentialPool { get; } = new();
+		private CredentialPool VsizeCredentialPool { get; } = new();
+		private Round Round { get; }
+		public IArenaRequestHandler ArenaRequestHandler { get; }
+		private BitcoinSecret BitcoinSecret { get; }
+		public KeyManager Keymanager { get; }
+		private SecureRandom SecureRandom { get; }
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private Coin[] Coins { get; set; }
+
+		public CoinJoinClient(
+			Round round,
+			IArenaRequestHandler arenaRequestHandler,
+			BitcoinSecret bitcoinSecret,
```

Comment:
We need one `BitcoinSecret` for each `Coin`, except for those that share the same script (address reused). So, I think you could have something like `IEnumerable<(Coin, BitcoinSecret)>` or at least that should be enough for this very first version.

Reply:
Can we use `keyManager`?

Reply:
How about this? https://github.com/molnard/WalletWasabi/blob/467d76b0e07c08565cdda10669fa8a0212633715/WalletWasabi/WabiSabi/Client/CoinJoinClient.cs#L94

Reply:
That makes sense to me, and to avoid threading issues with `KeyManager` it can be done before the client parallelism in a single threaded way.

Reply:
ohhh sorry, my mistake. there is only one secret by alice.